### PR TITLE
Uses forked pkg/util/interner for dogstatsd workload

### PR DESCRIPTION
**Pull request by scottopell**
### What does this PR do?

Replaces the dogstatsd string interner with a gc-friendlier version that doesn't require dumping the cache.

### Motivation

Testing out differences in string interning.

### Additional Notes

This will not be merged as-is as it copy-pastas most of `pkg/util/intern` instead of using it directly.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by blt** [comment]
I've introduced #20036 to give us a broader range of experimental throughput. This analysis will concern itself with run-id `0e3112e1-e30a-4211-9b28-915b909d5219`.

| Experiment | CPU time | Lifetime Allocations | Heap Live Size | Mutex Time | 
| ------------ | --------- | ------------------- | --------------- | ------------ |
| `uds_dogstatsd_to_api_nodist_1KiB` | ~0 | ~ -15k | 0 Mb | 0ms |
| `uds_dogstatsd_to_api_nodist_100MiB` | -27sec | -50M | +30MB | ~ -1sec |
| `uds_dogstatsd_to_api_nodist_200MiB` | -28sec | -60M | -106Mb | +4sec | 

Compare to the [raising of the default size](https://github.com/DataDog/datadog-agent/pull/19882#issuecomment-1743602854) this PR sees a similar reduction in lifetime allocations and a curious rise in heap live size in only one experiment, otherwise the drop in `_200MiB` heap live size is improved. If we consider [capture data](https://app.datadoghq.com/dashboard/t3q-y5d-i7u/smp-regression-detector-capture-data?refresh_mode=paused&tpl_var_experiment%5B0%5D=uds_dogstatsd_to_api_nodist_100mib&tpl_var_experiment%5B1%5D=uds_dogstatsd_to_api_nodist_1kib&tpl_var_experiment%5B2%5D=uds_dogstatsd_to_api_nodist_200mib&tpl_var_run-id%5B0%5D=0e3112e1-e30a-4211-9b28-915b909d5219&from_ts=1696958617936&to_ts=1696959179648&live=false) throughput is improved below threshold in the `_200MiB` case while in the `_100Mib` it remains constant as before, implying that although this change does allow more throughput through the Agent's throughput in these experiments is stable <= 100MiB but becomes unstable somewhere around 120MiB/sec, 125MiB/sec if this PR is merged. CPU utilization is improved in a manner similar to the default size raise PR. 

I would like to see this run repeated with #20036 in its history. I see the results here as positive, an improvement over `main`.

**Comment by blt** [comment]
This analysis will concern itself with run-id `96f38e65-f7ae-4073-8fab-dbb569b4e089`.

| Experiment | CPU time | Lifetime Allocations | Heap Live Size | Mutex Time | 
| ------------ | --------- | ------------------- | --------------- | ------------ |
| `uds_dogstatsd_to_api_nodist_1KiB` | ~0 | ~ 0 | 0 Mb | - 8ms |
| `uds_dogstatsd_to_api_nodist_1MiB` | ~0 | ~ -100k | -2 Mb | ~0ms |
| `uds_dogstatsd_to_api_nodist_16MiB` |  -2 sec | - 8M | +18 Mb | +100ms |
| `uds_dogstatsd_to_api_nodist_32MiB` | -3 sec | - 16M | -11 Mb | + 1 sec |
| `uds_dogstatsd_to_api_nodist_100MiB` | -29 sec | -49M | -59Mb | +1 sec |
| `uds_dogstatsd_to_api_nodist_200MiB` | -30 sec | -59M | -146Mb | +1 sec |

A consistent result. The low end is not pessimized and the high end is improved. Looking at the 16MiB case [capture data](https://app.datadoghq.com/dashboard/t3q-y5d-i7u/smp-regression-detector-capture-data?refresh_mode=paused&tpl_var_experiment%5B0%5D=uds_dogstatsd_to_api_nodist_16mib&tpl_var_run-id%5B0%5D=96f38e65-f7ae-4073-8fab-dbb569b4e089&from_ts=1696973440763&to_ts=1696973972931&live=false) does bear out the higher memory consumption. Throughput is improved, [profile](https://app.datadoghq.com/profiling/comparison?query=account%3Asingle-machine-performance%20runtime%3Ago%20client_team%3Aagent%20version%3A%28_7_50_0_devel_git.49.e4fe597_e4fe597_%20OR%20_7_50_0_devel_git.57.3185ebc_3185ebc_%29&compare_end_A=1696991320754&compare_end_B=1696991612148&compare_query_A=account%3Asingle-machine-performance%20runtime%3Ago%20client_team%3Aagent%20version%3A%28_7_50_0_devel_git.49.e4fe597_e4fe597_%29%20experiment%3Auds_dogstatsd_to_api_nodist_16mib%20&compare_query_B=account%3Asingle-machine-performance%20runtime%3Ago%20client_team%3Aagent%20version%3A%28_7_50_0_devel_git.57.3185ebc_3185ebc_%29%20experiment%3Auds_dogstatsd_to_api_nodist_16mib%20&compare_start_A=1696971198955&compare_start_B=1696970694470&compareValuesMode=absolute&my_code=disabled&profile_type=heap-live-size&viz=flame_graph&start=1696956697961&end=1697043097961&paused=false) suggests the bulk of the greater heap is in the UDS listener and then partially in the string interner. I don't have a hypothesis for that result.

**Comment by remeh** [comment]
@blt hijacking a bit the thread, sorry, but for a related question.

I was thinking about the `Heap Live Size` which appears to be slightly higher in one scenario [in current PR](https://github.com/DataDog/datadog-agent/pull/19990#issuecomment-1758131870) (your comment right above) but also in [in this comment](https://github.com/DataDog/datadog-agent/pull/19882#issuecomment-1743602854) in #19882 increasing the string interner size.

AFAIR, both implementations rely on the GC to use the equivalent of weak pointers/references, where the GC could possible collect them and that's OK. Wouldn't it mean that the moments the GC runs has an impact on the `Heap Live Size` result, that is, the moment the data is collected also has an impact? If this value is often sampled at a time where the GC hasn't intervened, you would possibly see differences between scenarios. Should we consider scenarios where the GC is manually kicked off before collecting samples?

**Comment by blt** [comment]
> @blt hijacking a bit the thread, sorry, but for a related question.

It's an interesting question and apropos I think.
 
> I was thinking about the `Heap Live Size` which appears to be slightly higher in one scenario [in current PR](https://github.com/DataDog/datadog-agent/pull/19990#issuecomment-1758131870) (your comment right above) but also in [in this comment](https://github.com/DataDog/datadog-agent/pull/19882#issuecomment-1743602854) in #19882 increasing the string interner size.
> 
> AFAIR, both implementations rely on the GC to use the equivalent of weak pointers/references, where the GC could possible collect them and that's OK. Wouldn't it mean that the moments the GC runs has an impact on the `Heap Live Size` result, that is, the moment the data is collected also has an impact? 

It would have an impact, yeah. 

> If this value is often sampled at a time where the GC hasn't intervened, you would possibly see differences between scenarios. Should we consider scenarios where the GC is manually kicked off before collecting samples?

I'd be vaguely against that. Which is to say, I think the profiler is not quite the right tool for this and I could have called this out a little better. Because of the sampling problem, like you point out, we might sample the Agent before GC has kicked on but that _is_ memory use that the user of Agent would actually see, so should we elide it by manually triggering GC before sampling? I think that would give us a different view, but it may be a skewed view. The right tool here is to examine the [capture data](https://app.datadoghq.com/dashboard/t3q-y5d-i7u/smp-regression-detector-capture-data?fullscreen_end_ts=1696973972931&fullscreen_paused=true&fullscreen_refresh_mode=paused&fullscreen_section=overview&fullscreen_start_ts=1696973440763&fullscreen_widget=7196294862350437&refresh_mode=paused&tpl_var_experiment%5B0%5D=uds_dogstatsd_to_api_nodist_16mib&tpl_var_run-id%5B0%5D=96f38e65-f7ae-4073-8fab-dbb569b4e089&from_ts=1696973440763&to_ts=1696973972931&live=false) which is itself sampled but at 1hz. 

I guess my argument would be that a freshly GC'ed Agent isn't any less representative than one where the GC has yet to run. Both states would be observable to an Agent user and need thought about when deciding resource allocation etc.

**Comment by blt** [comment]
Which is an interesting alternative view. Check this out:

* [max RSS per experiment](https://app.datadoghq.com/dashboard/t3q-y5d-i7u/smp-regression-detector-capture-data?fullscreen_end_ts=1696973972931&fullscreen_paused=true&fullscreen_refresh_mode=paused&fullscreen_section=split%20graph&fullscreen_start_ts=1696973440763&fullscreen_widget=1476210537773866&refresh_mode=paused&tpl_var_experiment%5B0%5D=uds_dogstatsd_to_api_nodist_32mib&tpl_var_experiment%5B1%5D=uds_dogstatsd_to_api_nodist_1mib&tpl_var_experiment%5B2%5D=uds_dogstatsd_to_api_nodist_16mib&tpl_var_experiment%5B3%5D=uds_dogstatsd_to_api_nodist_100mib&tpl_var_experiment%5B4%5D=uds_dogstatsd_to_api_nodist_64mib&tpl_var_experiment%5B5%5D=uds_dogstatsd_to_api_nodist_200mib&tpl_var_run-id%5B0%5D=96f38e65-f7ae-4073-8fab-dbb569b4e089&from_ts=1696973440763&to_ts=1696973972931&live=false)
* [max CPU per experiment](https://app.datadoghq.com/dashboard/t3q-y5d-i7u/smp-regression-detector-capture-data?fullscreen_end_ts=1696973972931&fullscreen_paused=true&fullscreen_refresh_mode=paused&fullscreen_section=split%20graph&fullscreen_start_ts=1696973440763&fullscreen_widget=3555058649880600&refresh_mode=paused&tpl_var_experiment%5B0%5D=uds_dogstatsd_to_api_nodist_32mib&tpl_var_experiment%5B1%5D=uds_dogstatsd_to_api_nodist_1mib&tpl_var_experiment%5B2%5D=uds_dogstatsd_to_api_nodist_16mib&tpl_var_experiment%5B3%5D=uds_dogstatsd_to_api_nodist_100mib&tpl_var_experiment%5B4%5D=uds_dogstatsd_to_api_nodist_64mib&tpl_var_experiment%5B5%5D=uds_dogstatsd_to_api_nodist_200mib&tpl_var_run-id%5B0%5D=96f38e65-f7ae-4073-8fab-dbb569b4e089&from_ts=1696973440763&to_ts=1696973972931&live=false)

The CPU numbers with the new interner are consistently on-par or better especially as throughput increases but the memory consumption data is less rosy on the low-end. It might be that with the way sampling works the live heap size ought to be dropped out of consideration, and only referenced from capture data.

**Comment by scottopell** [comment]
IIRC the GC is run before the heap state is captured for the profiler. 
Along the same lines, a concern of periodic tasks (ie, flush intervals, garbage collections) giving an un-representative view is what eventually led to https://github.com/DataDog/datadog-agent/pull/20081 . I believe this is more of a concern for cpu profiles compared to heap profiles due to the way heap profiling works in go (continuously accumulated) vs cpu profiling (run on-demand)

I'll rebase and kick off another run of this to include the changes from #20081

File: comp/dogstatsd/server/intern.go [hunk 1]
```diff
@@ -6,93 +6,99 @@
...
 var (
 	// There are multiple instances of the interner, one per worker. Counters are normally fine,
 	// gauges require special care to make sense. We don't need to clean up when an instance is
 	// dropped, because it only happens on agent shutdown.
-	tlmSIResets = telemetry.NewSimpleCounter("dogstatsd", "string_interner_resets",
-		"Amount of resets of the string interner used in dogstatsd")
-	tlmSIRSize = telemetry.NewSimpleGauge("dogstatsd", "string_interner_entries",
-		"Number of entries in the string interner")
-	tlmSIRBytes = telemetry.NewSimpleGauge("dogstatsd", "string_interner_bytes",
-		"Number of bytes stored in the string interner")
```
Comment (iksaif) [inline comp/dogstatsd/server/intern.go:23]: This is really really valuable imformation, can we try to keep it?

**Comment by scottopell** [comment]
Exploring a slightly different architecture in https://github.com/DataDog/datadog-agent/pull/20335

File: comp/dogstatsd/server/intern.go (hunk 1 above)
Comment (scottopell) [reply comp/dogstatsd/server/intern.go:23]: yup, don't worry I'll make sure all existing telemetry is covered in a final version of this PR.

**Pull request closed without merging**